src/ansi-c/padding.cpp

```diff
@@ -48,25 +50,26 @@ mp_integer alignment(const typet &type, const namespacet &ns)
   }
 
   // alignment but no packing
-  if(a_int>0 && !type.get_bool(ID_C_packed))
+  if(a_int > 0 && !type.get_bool(ID_C_packed) && !in_packed)
     return a_int;
   // no alignment, packing
-  else if(a_int==0 && type.get_bool(ID_C_packed))
+  else if(a_int == 0 && (type.get_bool(ID_C_packed) || in_packed))
     return 1;
 
   // compute default
   mp_integer result;
 
   if(type.id()==ID_array)
-    result = alignment(to_array_type(type).element_type(), ns);
+    result = alignment(to_array_type(type).element_type(), ns, in_packed);
   else if(type.id()==ID_struct || type.id()==ID_union)
   {
+    const bool packed = in_packed || type.get_bool(ID_C_packed);
     result=1;
 
     // get the max
     // (should really be the smallest common denominator)
     for(const auto &c : to_struct_union_type(type).components())
-      result = std::max(result, alignment(c.type(), ns));
+      result = std::max(result, alignment(c.type(), ns, packed));
```

Comment:
Does packing of an outer struct really affect the inner one?

Reply:
That's how I'd interpret GCC's docs, yes: from https://gcc.gnu.org/onlinedocs/gcc/Common-Type-Attributes.html#Common-Type-Attributes (note the second sentence): 
> This attribute, attached to a struct, union, or C++ class type definition, specifies that each of its members (other than zero-width bit-fields) is placed to minimize the memory required. This is equivalent to specifying the packed attribute on each of the members.

Reply:
Not sure that "nested members" count as "members" here?

Reply:
https://stackoverflow.com/questions/16904613/whats-the-effect-of-attribute-packed-on-nested-structs

Reply:
I've got more work to do it seems: I have expanded the regression test, and it is now failing (GCC things the static assertions are fine).